src/Aspire.Hosting/Dcp/ApplicationExecutor.cs

```diff
@@ -599,58 +556,83 @@ string CombineUrls(string url, string launchUrl)
+                var ep = resourceWithEndpoints.GetEndpoint(endpointName);
+
+                if (ep.EndpointAnnotation.FromLaunchProfile &&
+                    appModelResource is ProjectResource p &&
+                    p.GetEffectiveLaunchProfile() is LaunchProfile profile &&
+                    profile.LaunchUrl is string launchUrl)
+                {
+                    // Concat the launch url from the launch profile to the urls with IsFromLaunchProfile set to true
+
+                    string CombineUrls(string url, string launchUrl)
+                    {
+                        if (!launchUrl.Contains("://"))
+                        {
+                            // This is relative URL
+                            url += $"/{launchUrl}";
+                        }
+                        else
+                        {
+                            // For absolute URL we need to update the port value if possible
+                            if (profile.ApplicationUrl is string applicationUrl
+                                && launchUrl.StartsWith(applicationUrl))
+                            {
+                                url = launchUrl.Replace(applicationUrl, url);
+                            }
+                        }
+
+                        return url;
+                    }
 
-                var endpointString = $"{uriScheme}://{endpoint.Spec.Address}:{endpoint.Spec.Port}";
-                var proxyUrlString = $"{uriScheme}://{service.AllocatedAddress}:{service.AllocatedPort}";
-                var isFromLaunchProfile = false;
+                    if (ep.IsAllocated)
+                    {
+                        var url = CombineUrls(ep.Url, launchUrl);
 
-                if (launchProfile is not null)
+                        urls.Add(new(ep.EndpointName, url, false));
+                    }
+                }
+                else
                 {
-                    _ = bool.TryParse(launchProfile, out isFromLaunchProfile);
+                    if (ep.IsAllocated)
+                    {
+                        urls.Add(new(ep.EndpointName, ep.Url, false));
+                    }
                 }
 
-                endpoints.Add(new(endpointString, proxyUrlString, isFromLaunchProfile));
-            }
-        }
-
-        if (_resourceAssociatedServicesMap.TryGetValue((resourceKind, name), out var resourceServiceMappings))
-        {
-            foreach (var serviceName in resourceServiceMappings)
-            {
-                if (_servicesMap.TryGetValue(serviceName, out var service))
+                if (ep.EndpointAnnotation.IsProxied)
                 {
-                    services.Add(new(service.Metadata.Name, service.AllocatedAddress, service.AllocatedPort));
+                    var endpointString = $"{ep.Scheme}://{endpoint.Spec.Address}:{endpoint.Spec.Port}";
+                    urls.Add(new($"{ep.EndpointName}-listen-port", endpointString, true));
```

Comment:
nit: consider naming literals argument for context:

```suggestion
                    urls.Add(new(Name: $"{ep.EndpointName}-listen-port", endpointString, IsInternal: true));
```

Reply:
Done